### PR TITLE
Blackbox

### PR DESCRIPTION
**Pull request by CaesarQ**
New functionality for attacks on blackboxes.  Including:

- Gradient free optimizer based attacks (Evolutionary Algorithms, Bandits)
- Gradient estimation attacks (Finite Difference, NES)
- Wrappers for gradient approximation.  Allows any model to be used with any attack in advertorch.

### CODE REVIEW

File: advertorch/attacks/blackbox/bandits.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            fd_eta=0.01, exploration=0.01, online_lr=0.1,
+            loss_fn=None, 
+            nb_iter=40,
+            eps_iter=0.01,
+            clip_min=0., clip_max=1.,
+            targeted : bool = False
+            ):
+
+        if loss_fn is not None:
+            import warnings
```
Comment (msalihs) [inline advertorch/attacks/blackbox/bandits.py:138]: Imports on top of the file

File: advertorch/attacks/blackbox/bandits.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            exploration=self.exploration, online_lr=self.online_lr, 
+            nb_iter=self.nb_iter, eps_iter=self.eps_iter
+        )
+
+        adv = adv.reshape(shape)
+        
+        return adv
+        
+
+
```
Comment (msalihs) [inline advertorch/attacks/blackbox/bandits.py:206]: Single new line at the end of the file.

File: advertorch/attacks/fast_adaptive_boundary.py [hunk 3]
```diff
@@ -13,6 +13,15 @@
 import torch
 import time
 
+def zero_gradients(x):
```
Comment (msalihs) [inline advertorch/attacks/fast_adaptive_boundary.py:16]: Put the function definition after all the imports

**Review by msalihs — COMMENTED** [review]
Just a few comments regarding style

File: advertorch/attacks/blackbox/bandits.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    :param prior_init: (default None)
+    :param fd_eta: step-size used for finite difference grad estimate (default 0.01)
+    :param exploration: scales the exploration around prior (default 0.01)
+    :param online_lr: learning rate for the prior (default 0.1)
+    :param nb_iter: number of iterations (default 40)
+    :param eps_iter: attack step size (default 0.01)
+
+    :return: tuple of tensors containing (1) the adv example, (2) the prior
+    """
+    #if so, we could just use the same estimate grad
```
[the remark this answers is not in the document]
Comment (CaesarQ) [reply advertorch/attacks/blackbox/bandits.py:46]: It was preceded by a comment which was moved.  I have now moved this snippet to the docstrings of another class.

File: advertorch/attacks/blackbox/bandits.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    else:
+        prior = prior_init.clone()
+
+    for t in range(nb_iter):
+        #before: # [nbatch, ndim, nsamples]
+        #now: # [nbatch, ndim]
+        exp_noise = exploration * torch.randn_like(prior)/(ndim**0.5)
+        
+        # Query deltas for finite difference estimator
+        ##...this step needs to change
```
[the remark this answers is not in the document]
Comment (CaesarQ) [reply advertorch/attacks/blackbox/bandits.py:68]: No, it was already previously address, but the comment remained unchanged.  Updated.

File: advertorch/attacks/blackbox/bandits.py (hunk 1 above)
Comment (CaesarQ) [reply advertorch/attacks/blackbox/bandits.py:138]: Fixed.
